### ISSUE TITLE
Relative paths are treated relative to config file instead of cwd

### ISSUE THREAD

**Issue by RikkiGibson** [Azure/autorest#2519]
When I invoke AutoRest like so:

```sh
cd /mycwd
node /path/to/autorest/src/autorest-core/dist/app.js /path/to/config --output-folder=relative/output/path
```
the outputs are written to `/path/to/config/relative/output/path` instead of `/mycwd/relative/output/path`.

Is this by design, or would it make sense to resolving relative paths given from the command line relative to the current working directory, as people giving relative paths in the shell may expect?

Also, passing an output path containing a ~, for instance, causes a folder named ~ to be created, so maybe resolving paths passed on the command line would just be generally helpful.

**Comment by olydis** [comment]
This is by design, at least right now. Reasoning: 100% reproducibility. For consistencies sake, the CLI and configuration files are treated absolutely identically, i.e. one can always move any parameter from the CLI to the config file or vice versa. The config file is the preferred spot for parameters, so it "weighs more" in our decision to make it the implicit `base-folder` (which is the thing *everything* that's relative is resolved against).
So when adding an `input-file` (where relative to config file usually makes things easy) on the CLI, you can specify it relatively instead of having to suddenly copy paste the path to the config file. Only for output paths, relative behavior is arguably unintuitive.

So currently everything is perfectly reproducible, i.e. same config and parameters, exactly same behavior, regardless of CWD. I guess it's a trade-off, open for discussion!

**Comment by RikkiGibson** [comment]
I can see that you guys had to make a choice between options with different pros and cons. I think that there are a few things that can be done to make it less confusing for users.

1. We often refer to config files over HTTP. If a relative output-folder is given, it's relative to that URL and there's nothing reasonable that AutoRest can do to honor this configuration (is there?). From playing with it, AutoRest appears to silently write outputs nowhere in this case. It should probably show a validation error and exit immediately.

2. I suspect that output-folder is a special case for configuration where users very rarely intend to specify it relative to the config file. It may make sense to log INFORMATION if output-folder is specified with a relative path.

My experience getting up to speed on this behavior was spending most of the morning one day hooking up a debugger to AutoRest and trying different versions of it to figure out why it would be getting so far through code generation to fail just when writing to the file system. I eventually figured out it because I happened to start using a local specs repo and I decided to try running opensnoop while AutoRest ran to inspect *all* file I/O on the system.
